rs/src/parse/relations/join.rs

```diff
@@ -106,59 +122,81 @@ pub fn parse_join_rel(x: &substrait::JoinRel, y: &mut context::Context) -> diagn
-        y.push_summary(
-            comment::Comment::new()
-                .nl()
-                .plain(format!("The result is filtered by {filter_expression:#}.")),
-        );
+    let nullable = if join_expression_node.data_type().nullable() {
+        "false or null"
+    } else {
+        "false"
+    };
+    y.push_summary(
+        comment::Comment::new().nl().plain(match join_type {
+            JoinType::Unspecified => "".to_string(),
+            JoinType::Inner => format!(
+                " 
+Returns rows combining the row from the left and right
+input for each pair where the join expression yields true.
+Discarding rows where the join expression yields {}.",
+                nullable
+            ),
+            JoinType::Outer => format!(
+                " 
+Returns rows combining the row from the left and right
+input for each pair where the join expression yields true.
+Discarding rows where the join expression yields {}.
+If the join expression never yields true for any left or
+right row, this returns a row anyway, with the fields
+corresponding to the other input set to null.",
+                nullable
+            ),
+            JoinType::Left => format!(
+                " 
+Returns rows combining the row from the left and right
+input for each pair where the join expression yields true.
+Discarding rows where the join expression yields {}.
+If the join expression never yields true for a row from the
+left, this returns a row anyway, with the fields corresponding
+to the right input set to null.",
+                nullable
+            ),
+            JoinType::Right => format!(
+                " 
+Returns rows combining the row from the left and right
+input for each pair where the join expression yields true.
+Discarding rows where the join expression yields {}.
+If the join expression never yields true for a row from the
+right, this returns a row anyway, with the fields corresponding
+to the left input set to null.",
+                nullable
+            ),
+            JoinType::Semi => " 
+Filters rows from the left input, propagating a row only if
+the join expression yields true for that row combined with
+any row from the right input."
+                .to_string(),
+            JoinType::Anti => " 
+Filters rows from the left input, propagating a row only if
+the join expression does not yield true for that row combined
+with any row from the right input."
+                .to_string(),
+            JoinType::Single => " 
+Returns a row for each row from the left input, concatenating
+it with the row from the right input for which the join
+expression yields true. If the expression never yields true for
+a left input, the fields corresponding to the right input are
+set to null. If the expression yields true for a left row and
+multiple right rows, this may return the first pair encountered
+or throw an error."
+                .to_string(),
+        }),
+    );
```

Comment:
I don't really like this string style. For one thing, I hate multiline strings that break the indentation flow of the code, but I'm also not 100% sure what happens or should happen with newlines in these strings currently. I guess they don't break anything in HTML and aren't used in the plantext output because they're not diagnostics, but via protobuf export you could probably get the original strings back. I'd rather have them not put arbitrarily-placed newlines in them.

I guess the `concat!` thing can be replaced with a properly indented multiline string by just escaping the newlines, though? When I wrote it I probably wasn't aware or sure that Rust also removes indentation following an escaped newline (making the newline escape actually useful; looking at you, Python).

Reply:
How about [c463c86](https://github.com/substrait-io/substrait-validator/pull/52/commits/c463c864c2826614daa7c3c15d4f820af977045d)?

Reply:
I normally align the string contents to the same column as the first `"` rather than the first character of the string (and don't think cargo fmt has an opinion about this, unfortunately), but I'm fine with this.